data/field_type.go

```diff
@@ -0,0 +1,523 @@
+
+	case FieldTypeTime, FieldTypeNullableTime:
+		return FieldTypeNullableTime
+	default:
+		panic(fmt.Sprintf("unsupported vector ptype: %+v", p))
+	}
+}
+
+// NonNullableType converts the FieldType to the corresponding not-nullable type.
+func (p FieldType) NonNullableType() FieldType {
+	switch p {
+	// ints
+	case FieldTypeInt8, FieldTypeNullableInt8:
+		return FieldTypeInt8
+
+	case FieldTypeInt16, FieldTypeNullableInt16:
+		return FieldTypeInt16
+
+	case FieldTypeInt32, FieldTypeNullableInt32:
+		return FieldTypeInt32
+
+	case FieldTypeInt64, FieldTypeNullableInt64:
+		return FieldTypeInt64
+
+	// uints
+	case FieldTypeUint8, FieldTypeNullableUint8:
+		return FieldTypeUint8
+
+	case FieldTypeUint16, FieldTypeNullableUint16:
+		return FieldTypeUint16
+
+	case FieldTypeUint32, FieldTypeNullableUint32:
+		return FieldTypeUint32
+
+	case FieldTypeUint64, FieldTypeNullableUint64:
+		return FieldTypeUint64
+
+	// floats
+	case FieldTypeFloat32, FieldTypeNullableFloat32:
+		return FieldTypeFloat32
+
+	case FieldTypeFloat64, FieldTypeNullableFloat64:
+		return FieldTypeFloat64
+
+	// other
+	case FieldTypeString, FieldTypeNullableString:
+		return FieldTypeString
+
+	case FieldTypeBool, FieldTypeNullableBool:
+		return FieldTypeBool
+
+	case FieldTypeTime, FieldTypeNullableTime:
+		return FieldTypeTime
+	default:
+		panic(fmt.Sprintf("unsupported vector ptype: %+v", p))
+	}
+}
+
+// FieldTypeFromItemTypeString returns a field type from the current string
+//nolint:goconst,gocyclo
+func FieldTypeFromItemTypeString(s string) (FieldType, bool) {
+	switch s {
+	case "int8":
+		return FieldTypeInt8, true
+	case "*int8":
+		return FieldTypeNullableInt8, true
+
+	case "int16":
+		return FieldTypeInt16, true
+	case "*int16":
+		return FieldTypeNullableInt16, true
+
+	case "int32":
+		return FieldTypeInt32, true
+	case "*int32":
+		return FieldTypeNullableInt32, true
+
+	case "int":
+		fallthrough
+	case "int64":
+		return FieldTypeInt64, true
+	case "*int64":
+		return FieldTypeNullableInt64, true
+
+	case "unit8":
+		return FieldTypeUint8, true
+	case "*uint8":
+		return FieldTypeNullableUint8, true
+
+	case "uint16":
+		return FieldTypeUint16, true
+	case "*uint16":
+		return FieldTypeNullableUint16, true
+
+	case "uint32":
+		return FieldTypeUint32, true
+	case "*uint32":
+		return FieldTypeNullableUint32, true
+
+	case "uint64":
+		return FieldTypeUint64, true
+	case "*uint64":
+		return FieldTypeNullableUint64, true
+
+	case "float32":
+		return FieldTypeFloat32, true
+	case "*float32":
+		return FieldTypeNullableFloat32, true
+
+	case "double":
+		fallthrough
+	case "float":
+		fallthrough
+	case "float64":
+		return FieldTypeFloat64, true
+	case "*float64":
+		return FieldTypeNullableFloat64, true
+
+	case "string":
+		return FieldTypeString, true
+	case "*string":
+		return FieldTypeNullableString, true
+
+	case "bool":
+		return FieldTypeBool, true
+	case "*bool":
+		return FieldTypeNullableBool, true
+
+	case "time":
+		fallthrough
```

Comment:
maybe rewrite this using `case "time", "time.Time":` instead of `fallthrough`?

Reply:
i had no idea!  thanks